src/postgres/commands/executePostgresQuery.ts

```diff
@@ -0,0 +1,62 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { Client, QueryResult } from 'pg';
+import * as vscode from 'vscode';
+import { AzExtTreeItem, IActionContext } from 'vscode-azureextensionui';
+import { ext } from '../../extensionVariables';
+import { localize } from '../../utils/localize';
+import { PostgresDatabaseTreeItem } from '../tree/PostgresDatabaseTreeItem';
+import { configurePostgresFirewall } from './configurePostgresFirewall';
+import { enterPostgresCredentials } from './enterPostgresCredentials';
+import { loadPersistedPostgresDatabase } from './registerPostgresCommands';
+
+export async function executePostgresQuery(context: IActionContext, treeItem?: PostgresDatabaseTreeItem): Promise<void> {
+    await loadPersistedPostgresDatabase();
+
+    if (!treeItem) {
+        if (ext.connectedPostgresDB) {
+            treeItem = ext.connectedPostgresDB;
+        } else {
+            treeItem = <PostgresDatabaseTreeItem>await ext.tree.showTreeItemPicker(PostgresDatabaseTreeItem.contextValue, context);
+        }
+    }
+
+    if (!treeItem.clientConfig) {
+        let children: AzExtTreeItem[] = await treeItem.loadAllChildren(context);
+
+        while (children.length === 1) {
```

Comment:
All of this tree stuff is just to make sure they have the firewall and credentials set up, correct? We can make this more efficient (and more readable IMO) if we avoid the tree to do this. Instead of the references looking like this:

> executePostgresQuery.ts --> PostgresDatabaseTreeItem.ts --> cred/firewall stuff

Hopefully you can refactor so that it looks more like this:

> executePostgresQuery.ts --> cred/firewall stuff (in a util file maybe?) <-- PostgresDatabaseTreeItem.ts

I assume the util file would have functions like `isFirewallSetup`, `validateFirewall`, etc.

Reply:
Having the firewall and credentials set up was part of it but the main reason to load children here was to populate `clientConfig` so we could execute a query. I'll think more on how to populate that